source/fuzz/transformation_replace_add_sub_mul_with_carrying_extended.cpp

```diff
@@ -0,0 +1,194 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_replace_add_sub_mul_with_carrying_extended.h"
+
+#include "source/fuzz/fuzzer_util.h"
+
+namespace spvtools {
+namespace fuzz {
+
+namespace {
+const uint32_t kOpCompositeExtractIndexLowOrderBits = 0;
+const uint32_t kArithmeticInstructionIndexLeftInOperand = 0;
+const uint32_t kArithmeticInstructionIndexRightInOperand = 1;
+const uint32_t kOpTypeIndexSignedness = 2;
+}  // namespace
+
+TransformationReplaceAddSubMulWithCarryingExtended::
+    TransformationReplaceAddSubMulWithCarryingExtended(
+        const spvtools::fuzz::protobufs::
+            TransformationReplaceAddSubMulWithCarryingExtended& message)
+    : message_(message) {}
+
+TransformationReplaceAddSubMulWithCarryingExtended::
+    TransformationReplaceAddSubMulWithCarryingExtended(uint32_t struct_fresh_id,
+                                                       uint32_t struct_type_id,
+                                                       uint32_t result_id) {
+  message_.set_struct_fresh_id(struct_fresh_id);
+  message_.set_struct_type_id(struct_type_id);
+  message_.set_result_id(result_id);
+}
+
+bool TransformationReplaceAddSubMulWithCarryingExtended::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext&) const {
+  // |message_.struct_fresh_id| must be fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.struct_fresh_id())) {
+    return false;
+  }
+
+  // |message_.result_id| must refer to a suitable OpIAdd, OpISub or OpIMul
+  // instruction.
+  auto instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.result_id());
+  auto instruction_opcode = instruction->opcode();
+
+  switch (instruction_opcode) {
+    case SpvOpIAdd:
+    case SpvOpISub:
+      if (!TransformationReplaceAddSubMulWithCarryingExtended::
+              IsAddSubInstructionSuitable(ir_context, instruction))
+        return false;
+      break;
+    case SpvOpIMul:
+      if (!TransformationReplaceAddSubMulWithCarryingExtended::
+              IsMulInstructionSuitable(ir_context, instruction))
+        return false;
+      break;
+    default:
+      return false;
+  }
+  return true;
+}
+
+void TransformationReplaceAddSubMulWithCarryingExtended::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  // |message_.struct_fresh_id| must be fresh.
+  assert(fuzzerutil::IsFreshId(ir_context, message_.struct_fresh_id()) &&
+         "|message_.struct_fresh_id| must be fresh");
+
+  auto original_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.result_id());
+  auto original_instruction_opcode = original_instruction->opcode();
+  auto operand_type_id = original_instruction->type_id();
```

Comment:
Consider removing these, since `original_instruction->opcode()` and `original_instruction->type_id()` are fairly simple.